flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterTableAdd.java

```diff
@@ -0,0 +1,150 @@
+        this.withParen = withParen;
+        this.addedColumns = addedColumns;
+        this.watermark = sqlWatermark;
+        this.constraint = constraint;
+    }
+
+    /** Returns the column constraints plus the table constraints. */
+    public List<SqlTableConstraint> getFullConstraints() {
+        List<SqlTableConstraint> ret = new ArrayList<>();
+        this.addedColumns.getList().stream()
+                .map(SqlTableColumnPosition.class::cast)
+                .map(SqlTableColumnPosition::getColumn)
+                .forEach(
+                        column -> {
+                            if (column instanceof SqlTableColumn.SqlRegularColumn) {
+                                SqlTableColumn.SqlRegularColumn regularColumn =
+                                        (SqlTableColumn.SqlRegularColumn) column;
+                                regularColumn.getConstraint().map(ret::add);
+                            }
+                        });
+        ret.addAll(this.constraint);
+        return ret;
+    }
+
+    public SqlNodeList getColumns() {
+        return addedColumns;
+    }
+
+    public Optional<SqlWatermark> getWatermark() {
+        return Optional.ofNullable(watermark);
+    }
+
+    @Nonnull
+    @Override
+    public List<SqlNode> getOperandList() {
+        return ImmutableNullableList.of(
+                getTableName(),
+                addedColumns,
+                watermark,
+                new SqlNodeList(constraint, SqlParserPos.ZERO));
+    }
+
+    @Override
+    public void unparse(SqlWriter writer, int leftPrec, int rightPrec) {
+        super.unparse(writer, leftPrec, rightPrec);
+        writer.keyword("ADD");
+        // Distinguish whether the added column/watermark/constraint is in a paren
+        if (withParen) {
+            SqlWriter.Frame frame =
+                    writer.startList(SqlWriter.FrameTypeEnum.create("sds"), "(", ")");
+            for (SqlNode column : addedColumns.getList()) {
+                printIndent(writer);
+                column.unparse(writer, leftPrec, rightPrec);
+            }
+            if (constraint.size() > 0) {
+                for (SqlTableConstraint constraint : constraint) {
+                    printIndent(writer);
+                    constraint.unparse(writer, leftPrec, rightPrec);
+                }
+            }
+            if (watermark != null) {
+                printIndent(writer);
+                watermark.unparse(writer, leftPrec, rightPrec);
+            }
+
+            writer.newlineAndIndent();
+            writer.endList(frame);
+        } else {
+            if (addedColumns.getList().size() == 1) {
+                // add single column case
+                addedColumns.getList().get(0).unparse(writer, leftPrec, rightPrec);
+            } else if (watermark != null) {
+                // add watermark case
+                watermark.unparse(writer, leftPrec, rightPrec);
+            }
+            // TODO: add constraint case
```

Comment:
Is it still TODO ?

Reply:
Yes, current it is need, but I remove it firstly in this PR